native_c_extension/c_custom_to_s.c

```diff
@@ -0,0 +1,493 @@
+        break;
+    }
+}
+
+
+
+// This is the method we added to test out passing parameters
+// from https://silverhammermba.github.io/emberb/c/
+VALUE method_c_custom_to_s(VALUE self, VALUE first) {
+  VALUE ret;
+  int buffer_max = MAX_STRING_LENGTH * 2;
+  char buffer[buffer_max];
+  
+  switch (TYPE(first)) {
+  case T_NIL:
+    ret = rb_str_new_cstr("");
+    break;
+  case T_TRUE:
+    ret = rb_str_new_cstr("true");
+    break;
+  case T_FALSE:
+    ret = rb_str_new_cstr("false");
+    break;      
+  case T_FIXNUM: {
+    sprintf(buffer, "%ld", FIX2LONG(first));
+    ret = rb_str_new_cstr(buffer);
+    break;
+  }
+  case T_FLOAT: {
+    sprintf(buffer, "%.15g", RFLOAT_VALUE(first));
+    ret = rb_str_new_cstr(buffer);
+    break;
+  }
+  case T_SYMBOL: {
+    char *name = (char *) rb_id2name(SYM2ID(first));
+    int max_len = strlen(name);
+    int offset = 0;
+    // +1 for : +1 for NULL
+    method_c_custom_to_s_check_buffer_size(offset, max_len + 2, buffer_max);
+    snprintf(buffer, max_len + 2, ":%s", name);
+    ret = rb_str_new_cstr(buffer);
+    break;
+  }
+  case T_STRING: {
+    int max_len = RSTRING_LEN(first);
+    int remaining_characters = 0;
+    if (max_len > MAX_STRING_LENGTH) {
+      remaining_characters = max_len - MAX_STRING_LENGTH;
+      max_len = MAX_STRING_LENGTH;
+    }
+    max_len += 1; // +1 for NULL written by snprintf
+    method_c_custom_to_s_check_buffer_size(0, max_len, buffer_max);
+    // something's strange with StringValueCStr and StringValuePtr,
+    // because sprintf causes a buffer overflow but snprintf doesn't.
+    snprintf(buffer, max_len, "%s", StringValuePtr(first));
+
+    // -1 to write the first byte over the NULL added by snprintf
+    int offset = max_len - 1;
+
+    if (remaining_characters > 0) {
+      char buffer_small[128];
+      sprintf(&buffer_small[0], " (...%d more characters)", remaining_characters);
+      int buffer_small_len = strlen(buffer_small);
+
+      // +1 for NULL
+      method_c_custom_to_s_check_buffer_size(offset, buffer_small_len + 1, buffer_max);
+      snprintf(&buffer[offset], buffer_small_len + 1, "%s", buffer_small);
+      offset += buffer_small_len;
+    }
+
+    // this reports the error:
+    // in `generate': source sequence is illegal/malformed utf-8 (JSON::GeneratorError)
+    /* char buffer_utf8_encoded[offset * 4]; */
+    /* int counter = 0; */
+    /* while (counter < offset) { */
+    /*   unsigned int character = buffer[counter]; */
+    /*   counter += utf8_encode(character, (unsigned char *) &buffer_utf8_encoded[counter]); */
+    /* } */
+    /* ret = rb_utf8_str_new(buffer_utf8_encoded, counter); */
+
+    // this is 600%-900%x faster than calling Ruby but reports the error:
+    // in `generate': source sequence is illegal/malformed utf-8 (JSON::GeneratorError)
+    // ret = rb_utf8_str_new(buffer, offset);
+
+    // call Ruby function to utf8 encode instead of encode in C
+    VALUE string_unencoded = rb_str_new_cstr(buffer);
```

Comment:
Ruby strings can have nulls inside them and one of them could have ended in the buffer, so this won't always work.

Reply:
Switched to `rb_str_new` and passing the buffer size.